src/objects.c

```diff
@@ -201,28 +199,44 @@ void SET_TYPE_OBJ(Obj obj, Obj type)
         break;
 #endif
     case T_PREC:
+#ifdef HPCGAP
+        MEMBAR_WRITE();
+#endif
```

Comment:
@rbehrends based on the discussion we had earlier on Slack, I understood you as saying this `MEMBAR_WRITE` and several afterwards in function `SET_TYPE_OBJ`, could go. Could you please confirm or correct this?

Also, can't we likewise remove `ReadGuard` statements?

Reply:
I've looked at it again. The guards and memory barriers were put in under the assumption that the type could be changed concurrently. This should not normally happen, but if we (as we did back then for T_DATOBJ) put them in the public region, it is technically still possible. The guards can go regardless, but if we want to permit changing the type again, we may still need memory barriers.

Reply:
But in all the cases in `SET_TYPE_OBJ`, we end up writing to the object, and for that call `ADDR_OBJ` -- shouldn't that indeed automatically insert even a `WriteGuard` ? How can we then ever change anything concurrently in this function? Maybe my understanding of how guards are supposed to work simply is wrong?

Reply:
For that matter, I also don't understand how the handling of atomic objects works a few lines above this: sure, we take a hash lock for those objects. But regardless of this, shouldn't a guard violation be triggered the moment we invoke `ADDR_OBJ` (at least once guards work again, with PR #2845) ? 

Clearly I am missing something fundamental.

Reply:
The specific case we are looking at here is objects in the public region. Guard checks will always succeed for the public region. (This is also why atomic objects never trigger guard errors and the relevant code is actually bracketed with `#ifndef WARD_ENABLED` to excise even those checks for performance reasons.)

Now, with `T_DATOBJ` we have the problem that the code as it exists makes the objects public. Alternatively, we can make them readonly, and there is a setting for that (`ReadOnlyDatObjs`). However, that is something we had disabled by default, because it caused a slew of errors.

But as `T_DATOBJ` objects are now public to avoid these errors, they can in theory be modified again and concurrently by `SET_TYPE_DATOBJ()`, so memory barriers may be needed in case somebody does that.

For `T_COMOBJ` and `T_POSOBJ`, both guards or memory barriers should not be needed anymore. As I recall, this was an unrelated artifact caused by how type setting functions originally worked, bypassing guard checks, and with `unward`, these are now protected by guards by default.

Reply:
@rbehrends you say "the relevant code is actually bracketed with `#ifndef WARD_ENABLED`" -- well, there are calls to `ADDR_OBJ` for atomic objects just a few lines above the line we are discussing right now, and I see no `#ifndef WARD_ENABLED` there? Nor were there any in the old location of that code. And in `ADDR_OBJ` resp. `PTR_BAG` resp. the guard, there isn't any code  do anything special to handle atomic objects -- or is there? I tried to find such a check in PR #2845 but couldn't find anything. Did I miss it?

Reply:
I'm sorry, I meant to write "the relevant code in `src/hpc/aobjects.c`", because that's where the hot paths are to be found. Normally, a successful guard check on public objects is (by design) very cheap, as it only compares the region pointer to `NULL`. The cost is generally low enough to not justify polluting every location where an access of an atomic object happens with an `#ifndef WARD_ENABLED`, which is also potentially error-prone. But in `aobjects.c`, these accesses are all over the place, so they have been turned off for the performance-critical parts.